Modules/posixmodule.c

```
@@ -3277,6 +3418,322 @@ os_lstat_impl(PyObject *module, path_t *path, int dir_fd)
+    type = 'int'
+    converter = 'optional_bool_converter'
+[python start generated code]*/
+/*[python end generated code: output=da39a3ee5e6b4b0d input=47de85b300eeb19e]*/
+
+/*[clinic input]
+
+os.statx
+
+    path : path_t(allow_fd=True)
+        Path to be examined; can be string, bytes, a path-like object or
+        open-file-descriptor int.
+
+    mask: unsigned_int(bitwise=True)
+        A bitmask of STATX_* constants defining the requested information.
+
+    *
+
+    dir_fd : dir_fd = None
+        If not None, it should be a file descriptor open to a directory,
+        and path should be a relative string; path will then be relative to
+        that directory.
+
+    follow_symlinks: bool = True
+        If False, and the last element of the path is a symbolic link,
+        statx will examine the symbolic link itself instead of the file
+        the link points to.
+
+    sync: optional_bool(c_default='-1') = None
+        If True, statx will return up-to-date values, even if doing so is
+        expensive.  If False, statx will return cached values if possible.
+        If None, statx lets the operating system decide.
+
+Perform a statx system call on the given path.
+
+It's an error to use dir_fd or follow_symlinks when specifying path as
+  an open file descriptor.
+
+[clinic start generated code]*/
+
+static PyObject *
+os_statx_impl(PyObject *module, path_t *path, unsigned int mask, int dir_fd,
+              int follow_symlinks, int sync)
+/*[clinic end generated code: output=fe385235585f3d07 input=148c4fce440ca53a]*/
+{
+    if (path_and_dir_fd_invalid("statx", path, dir_fd) ||
+        dir_fd_and_fd_invalid("statx", dir_fd, path->fd) ||
+        fd_and_follow_symlinks_invalid("statx", path->fd, follow_symlinks))
+        return NULL;
+
+    int result;
+    /* Future bits may refer to members beyond the current size of struct
+       statx, so we need to mask them off to prevent memory corruption. */
+    mask &= _Py_STATX_KNOWN;
```

Should this error if the mask was changed by this code?

---

Hmm, maybe.  We'd only adjust the mask for Python code that specified the mask numerically instead of using the `os.STATX_*` constants.  That's probably only code that wants to use members Python 3.16 knows about but Python 3.15 doesn't.  `os.statx_result` doesn't provide Python with an interface to access members the interpreter doesn't know about, so while code could use the numeric value of `STATX_FOO`, accessing `stx_foo` would raise `AttributeError`.  And that's how current code copes with `os.stat_result`'s members varying across versions.  But I'm open to arguments that failing fast is better and version-crossing code should instead change their mask based on the constants available.